core/src/main/java/smile/classification/DecisionTree.java

```diff
@@ -540,8 +558,15 @@ public boolean findBestSplit() {
                 }
             } else {
 
+                final int[] constFeatures_;
+                synchronized (constFeatures) {
```

Comment:
`synchronized` is used to sync CPU caches and for ordering.

If even other thread replaces `this.constFeatures`, it might not be seen in the current thread.
Missing some `j` in constFeatures by contention is allowed because it's an optional optimization.

Reply:
This "synchronized" isn't necessary: at this point, no other thread can be accessing this.constFeatures. In fact, you don't need constFeatures_ at all, as there are no other threads accessing this.constFeatures during the time when you're using constFeatures_.

Once MulticoreExecutor.run() has completed, any modifications to this.constFeatures will be visible to the current thread: MulticoreExecutor invokes Future.get() which involves a lock acquisition and release, which means that all modifications that happen-before the acquisition are visible to this thread. The thread that did the work has to have placed the result in the Future, which means that everything it did happens-before the Future.get(), so if it changed this.constFeatures, it will be visible to this thread.

Reply:
I intended to use `synchronized` for CPU cache store buffer to flush and sync `constFeatures_` among threads. 

No need to be atomic paying synchronization costs but better to sync often.

Reply:
> Once MulticoreExecutor.run() has completed, any modifications to this.constFeatures will be visible to the current thread: MulticoreExecutor invokes Future.get() which involves a lock acquisition and release, which means that all modifications that happen-before the acquisition are visible to this thread. 

True. It's just a timing issue of synchronization in the middle of findBestSplit or in the end of findBestSplit.